pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/database/IdentifiersDao.scala

```diff
@@ -1,82 +1,145 @@
+import java.sql.{BatchUpdateException, Statement}
 
 import grizzled.slf4j.Logging
 import scalikejdbc._
 import uk.ac.wellcome.models.work.internal.SourceIdentifier
-import uk.ac.wellcome.platform.idminter.exceptions.IdMinterException
 import uk.ac.wellcome.platform.idminter.models.{Identifier, IdentifiersTable}
 
 import scala.concurrent.blocking
 import scala.util.Try
 
+object IdentifiersDao {
+  case class LookupResult(found: Map[SourceIdentifier, Identifier],
+                          notFound: List[SourceIdentifier])
+  case class InsertResult(succeeded: List[Identifier])
+
+  case class InsertError(failed: List[Identifier],
+                         e: Throwable,
+                         succeeded: List[Identifier])
+      extends Exception
+}
+
 class IdentifiersDao(db: DB, identifiers: IdentifiersTable) extends Logging {
+  import IdentifiersDao._
 
   implicit val session = AutoSession(db.settingsProvider)
 
-  /* An unidentified record from the transformer can give us a list of
-   * identifiers from the source systems, and an ontology type (e.g. "Work").
-   *
-   * This method looks for existing IDs that have matching ontology type and
-   * source identifiers.
-   */
-  def lookupId(
-    sourceIdentifier: SourceIdentifier
-  ): Try[Option[Identifier]] = Try {
-
-    val sourceSystem = sourceIdentifier.identifierType.id
-    val sourceId = sourceIdentifier.value
-
-    // TODO: handle gracefully, don't TryBackoff ad infinitum
-    blocking {
-      debug(s"Matching ($sourceIdentifier)")
-
-      val i = identifiers.i
-      val query = withSQL {
-        select
-          .from(identifiers as i)
-          .where
-          .eq(i.OntologyType, sourceIdentifier.ontologyType)
-          .and
-          .eq(i.SourceSystem, sourceSystem)
-          .and
-          .eq(i.SourceId, sourceId)
-
-      }.map(Identifier(i)).single
-      debug(s"Executing:'${query.statement}'")
-      query.apply()
+  def lookupIds(sourceIdentifiers: Seq[SourceIdentifier]): Try[LookupResult] = {
+    Try {
+      debug(s"Matching ($sourceIdentifiers)")
+      val sqlParametersToSourceIdentifier =
+        buildSqlQueryParameters(sourceIdentifiers)
+
+      blocking {
+
+        val i = identifiers.i
+        val query = withSQL {
+
+          select
+            .from(identifiers as i)
+            .where
+            .in(
+              (i.OntologyType, i.SourceSystem, i.SourceId),
+              sqlParametersToSourceIdentifier.keys.toList)
+
+        }.map(rs => {
+            val foundParameter = buildSqlParametersFromResult(i, rs)
+            sqlParametersToSourceIdentifier.get(foundParameter) match {
+              case Some(sourceIdentifier) =>
+                (sourceIdentifier, Identifier(i)(rs))
+              case None =>
+                // this should be impossible in practice
+                throw new RuntimeException(
+                  "The values returned by the query could not be matched to a sourceIdentifier")
```

Comment:
Minor: if I get this “impossible” error, I’m not going to know why. Include `foundParameter` in the error message to make it easier to debug.